signals/signal_exporter.py

```diff
@@ -0,0 +1,383 @@
+        """
+        self.subs = []
+        self.proc_name = process_name
+        self.runner_host = runner_host
+        self.pub_id = process_name + "-" + str(uuid.uuid4())
+        self.redis = redis.Redis(host=redis_host, port=redis_port, db=0)
+        self.init_listener = None
+        self.legal_events = None
+
+    def _sig_builder(
+        self, event: str, sample: int = -1, tag: str = None, metadata: Dict = None
+    ) -> Signal:
+        """
+        Build a signal data object based on exporter object fields,
+        as well as user-inputted fields. Returns the signal object.
+        """
+        sig = Signal(publisher_id=self.pub_id, process_name=self.proc_name, event=event)
+        sig.runner_host = self.runner_host if self.runner_host else sig.runner_host
+        sig.tag = tag if tag else sig.tag
+        sig.metadata = metadata if metadata else sig.metadata
+        sig.sample_no = sample if sample >= 0 else sig.sample_no
+        return sig
+
+    def _get_data_dict(self, response: Dict) -> Dict:
+        """
+        Returns response signal payload if a properly-formed
+        response is received. Otherwise return None.
+        """
+        if not "data" in response:
+            print("No data in this response message")
+            return None
+        try:
+            data = json.loads(response["data"])
+        except ValueError:
+            return None
+        if "tool_id" not in data or "publisher_id" not in data or "event" not in data:
+            print("Malformed response data found")
+            return None
+        return data
+
+    def _fetch_responders(self) -> None:
+        """
+        Start initialization response listener. Add tool_ids from proper
+        responses to the subscriber list.
+        """
+        subscriber = self.redis.pubsub(ignore_subscribe_messages=True)
+
+        def _init_handler(item) -> None:
+            data = self._get_data_dict(item)
+            if (
+                data
+                and data["event"] == "initialization"
+                and data["publisher_id"] == self.pub_id
+            ):
+                self.subs.append(data["tool_id"])
+
+        subscriber.subscribe(**{"event-signal-response": _init_handler})
+        self.init_listener = subscriber.run_in_thread()
+
+    def _check_subs(self, event: str) -> Tuple[Any, list]:
+        """
+        Listen for responses from all registered subscribers. Return
+        listener, as well as value based on responders' RAS codes.
+        """
+        if not self.subs:
+            return None, [0]
+
+        to_check = set(self.subs)
+        subscriber = self.redis.pubsub(ignore_subscribe_messages=True)
+        result_box = [0]
+
+        def _sub_handler(item: Dict) -> None:
+            data = self._get_data_dict(item)
+            if data and data["publisher_id"] == self.pub_id and data["event"] == event:
+                if "ras" in data:
+                    to_check.remove(data["tool_id"])
```

Comment:
This is going to raise a key error if the ``tool_id`` isn't in the ``to_check`` set. Do you think that we could add in a check like this:
```python
if data["tool_id"] not in to_check:
    print(f"WARNING: Got a response from tool '{data['tool_id']}' but it's not on the known subscribers list")
else:
    to_check.remove(data["tool_id"])
```

Let me know though if I'm not understanding this correctly and/or if this isn't needed.

Reply:
I think it actually handles that case by simply removing nothing, but your solution is still better as it significantly increases clarity.

Reply:
The question is: if an unsubscribed tool responds with a bad response, do we still wanna give the warning that a tool returned a bad resp? Or would that not technically count as a "sub failing"? Really just a matter of whether or not to indent line 197 (see "non-sub handling" commit)

Reply:
I wish that sets would just remove nothing if the item wasn't in there, but it does raise a ``KeyError``:

```
>>> {"one", "two"}.remove("three")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'three'
>>> 
```

As a result I think it would be a good idea to add a check for this. I see what you're saying though about handling unsubscribed tools. Do you think then that we could do something like this:

```python
if data["tool_id"] in to_check: 
    to_check.remove(data["tool_id"])
```

And then add a print statement to log which tools we got a response from to help with debug?

Reply:
Ok I completely misunderstood your last comment and also just realized I was looking at outdated code :( my bad. What you pushed looks great. IMHO I think indenting the block at line 209 so we don't fail on unsubscribed responses would be a good idea. I'll trust your judgement on that though.

Reply:
I agree, I think it doesn't make sense to fail on a non-sub. What I can do though is in the warning for the non-sub also display its ras, just so the info is available